### DISCUSSION TITLE
Is there a secondary development documentation here?

### DISCUSSION THREAD

**Discussion by yanggjjie998** [SCADA-LTS/Scada-LTS#2312] [General]
I want to add new data sources.
Is there documentation to help me redevelop.
Or if there is documentation detailing the structure of the project that tells me what needs to be changed to support the new data source.

**Comment by Limraj** [comment]
Hello,
you can connect to external databases by configuring the 'Data Source' type SQL in application (Data Sources Page) you can follow the configuration from tomcat_home/conf/context.xml.
example:
![Screenshot 2022-08-04 at 12 03 35](https://user-images.githubusercontent.com/35842300/182820986-2e767a64-08f1-49d6-9255-f7e9df62a302.png)

Import this config (Import/Export-> content file past to 'Data' -> Click "Import"
[sql_example.txt](https://github.com/SCADA-LTS/Scada-LTS/files/9258859/sql_example.txt)
![Screenshot 2022-08-04 at 12 27 25](https://user-images.githubusercontent.com/35842300/182825581-e49eb2ab-252e-4c57-960c-30ba9927ba2a.png)

**However, keep in mind that these connections are not taken from connection pools, the connection has its own DataSource object, which can impact performance. The mysql database should be set to the appropriate value in max_connections.**

Best wishes.

**Comment by Limraj** [comment]
I'am created issue: #2313

**Comment by yanggjjie998** [comment]
hello @Limraj 
I'm sorry, maybe I didn't express myself clearly, What I want to know is if there are any such documents like [doc](https://rapidscada.org/device-drivers-development/).
![p](https://user-images.githubusercontent.com/48661708/182982859-0e26eecc-c9e9-4f29-9d9b-9f7649deddd6.png)
I want to develop a data source driver myself, but I don't know what Java, JSP, JS files I need to add or what existing files I need to modify to support my new driver.
Anyway, the information you provided is of great help to my understanding of SCADA-LTS.

I have added the new driver JSP page and set up the jump, with the page content copied from Editsql.jsp. Did not change any code, but opened the page browser prompted me some errors, do I have missed something by doing this step?
![debuginfo](https://user-images.githubusercontent.com/48661708/182983498-458ee898-fb3a-47de-b924-3b558aa413ae.png)

Thank you very much.

**Reply by yanggjjie998** [reply]
com.serotonin.mango.MangoContextListener#constantsInitialize  add a new driver Attribute.
the NullPointerException has been resolved.

**Comment by Limraj** [comment]
Hello,
okay. There is an approach in Java that the documentation is the code itself. This is purely programming work. You would have to see how Data Source SQL is implemented, for example. So classes with the ending RT (runtime, logic, operation), VO (value object, data, serialization to blob/json) with the data of a given datasource, and its Data Points / Points Locator, analogically RT and VO classes. Old front use DWR technology and jsp, classes with ending DWR. (A new frontend based on vue.js and REST API is under development) Here is an example implementation of new Data Source (Tango): https://github.com/SCADA-LTS/Scada-LTS/pull/2230/files

If you are not a java programmer, you can commission our team to implement a new Data Source. This is, among others, our business model. 

In the future, we are also considering a more configurational approach to creating new Data Sources.
Out of curiosity, what Data Source do you need?

Best wishes.

**Reply by yanggjjie998** [reply]
Thank you for your reply.
After reading the source code,  have a basic understanding of how to do new data source driven development,
The Tango related information you provided is very helpful to me.
My plan is to carry out new data source driven development with the customized data exchange protocol within the company.
